### PR TITLE
WIP: Move API repository to protobuf from gogo

### PR DESCRIPTION
**Pull request by sdake**
This PR moves the API repository to protobuf and abandons gogo. As part of
the process I discovered several errors in the current Makefile.core.mk that
result in documentation either not being generated or not being generated
in a standard format.

I will likely pull these bug fixes out so that `make gen` works (it currently
fails) - while keeping gogo intact, until I can determine if gogo can be purged
from Istio. The GOGO dep offers very little value now, given the optimizations that
have been done to the proto tool.

### CODE REVIEW

**Comment by mandarjog** [comment]
@kyessenov PTAL

File: go.mod [hunk 1]
```diff
@@ -4,8 +4,8 @@ go 1.12
 
 require (
 	github.com/gogo/protobuf v1.3.1
-	github.com/golang/protobuf v1.3.5 // indirect
+	github.com/golang/protobuf v1.3.5
```
Comment (kyessenov) [inline go.mod:7]: You must use 1.4+, e.g. 1.4.2. Mixing two versions causes performance degradation.

**Comment by kyessenov** [comment]
I'm sure @costinm has some opinions on this.

**Comment by sdake** [comment]
> LGTM with a version bump. This will block any other API changes until istio is updated.

@kyessenov this work affects the following repos (so far...):
```
common-files
tools
istio
api
```

I don't plan to merge this work until I am certain everything is in working order with the new structure. As some of the work may require force merging, I'll wait until the full implementation is done before merging.

Thanks for providing early feedback.

FWW, I have moved tools and common-files to 1.4.3 (iirc) and will soon be updating API with that work.

Cheers,
-steve

**Comment by sdake** [comment]
> LGTM with a version bump. This will block any other API changes until istio is updated.

what version do you want to bump?

File: go.mod (hunk 1 above)
Comment (sdake) [reply go.mod:7]: ya - reflection also is not aviailable in 1.3.5, so 1.4 series released after march is a must.

**Comment by sdake** [comment]
Please note, I removed deepcopy and the gogo proto marshal files in a recent commit. I know deepcopy is needed for go_client - I am just simplifying my work a little temporarily. deepcopy will definitely be restored. The other feature - gogo marshal/unmarshal - is this needed?

Both of these files are generated by tools in the istio/toools/cmd, but they depend on gogo at the moment. They will need a rewrite. Fortunately they are very small.

Cheers,
-steve

**Comment by sdake** [comment]
For Reviewers:
```
The interesting part is in Makefile.core.mk. The other stuff is generated from this Makefile change as well as the introduction of new protos and protbuf 1.4.2.
```

**Comment by costinm** [comment]
Any reason we can't generate both ?

As I mentioned in the past, as long as K8S is using gogo I don't think we
can remove that dependency.

And it is horrible practice to force people who may depend on istio/api to
switch to proto.

**Comment by costinm** [comment]
Also - obviously this should not be merged until a PR to change istio/istio
is also almost ready.

**Comment by sdake** [comment]
@costin @howardjohn has written this excellent document which explains the rationale for moving to gogo: https://docs.google.com/document/d/1cyANUWgTP9K4UEmKoFQk_eDR2qTlgdrSC0NLD-HX1VU

While it does mot answer your question of whether both types of go bindings can be generated it seems clear that having  both types of bindings one based upon protobuf v1 +  gogo and another based upon protobuf b2 in the same code baed is asking for pain. We have suffered long enough within Istio with thix problem.

I can look into generate bindings for individuals that wish to stsy on protobuf v1 + gogo which is dead, although my primary motivation is moving us into thd future.

I defer to T&R or TOC around the topic of how Istio supports migration from protobuf
1 +gogo.

**Comment by sdake** [comment]
@costin @howardjohn has written this excellent document which explains the rationale for moving to gogo: https://docs.google.com/document/d/1cyANUWgTP9K4UEmKoFQk_eDR2qTlgdrSC0NLD-HX1VU

While it does mot answer your question of whether both types of go bindings can be generated it seems clear that having  both types of bindings one based upon protobuf v1 +  gogo and another based upon protobuf b2 in the same code baed is asking for pain. We have suffered long enough within Istio with thix problem.

I can look into generate bindings for individuals that wish to stsy on protobuf v1 + gogo which is dead, although my primary motivation is moving us into thd future.

I defer to T&R or TOC around the topic of how Istio supports migration from protobuf
1 +gogo.

**Comment by sdake** [comment]
@costin agree on the DNM status of this PR until thr stleast four repositories are ready for update for including istio/istio in a semi-atomic fashion.

cheers
steve

**Comment by costinm** [comment]
On Mon, Aug 24, 2020 at 7:42 PM Steven Dake <notifications@github.com>
wrote:

> @costin <https://github.com/costin> @howardjohn
> <https://github.com/howardjohn> has written this excellent document which
> explains the rationale for moving to gogo:
> https://docs.google.com/document/d/1cyANUWgTP9K4UEmKoFQk_eDR2qTlgdrSC0NLD-HX1VU
>

I think giving the impact - it would be great to get this approved, and
discussed in the TOC as well, since the impact is breaking
anyone who depended on the API ( if we are removing gogo generation)  ?


> While it does mot answer your question of whether both types of go
> bindings can be generated it seems clear that having both types of bindings
> one based upon protobuf v1 + gogo and another based upon protobuf b2 in the
> same code baed is asking for pain. We have suffered long enough within
> Istio with thix problem.
>

And we suffered because Envoy decided to do the same thing - pick protobuf
and not support gogo, forcing all people depending on Envoy APIs to
do the same choice.

K8S is still using gogo - and many other projects. Forcing others to follow
our choice is simply evil, and if envoy did it we should not follow the
same path and inflict pain on others.

I'm not saying we shouldn't use the protobuf version - in some cases it'll
help, but this is equivalent with an intentional breaking of an API.


> I can look into generate bindings for individuals that wish to stsy on
> protobuf v1 + gogo which is dead, although my primary motivation is moving
> us into thd future.
>

It'll be dead when K8S stops using it.

**Comment by sdake** [comment]
> On Mon, Aug 24, 2020 at 7:42 PM Steven Dake ***@***.***> wrote: @costin <https://github.com/costin> @howardjohn <https://github.com/howardjohn> has written this excellent document which explains the rationale for moving to gogo: https://docs.google.com/document/d/1cyANUWgTP9K4UEmKoFQk_eDR2qTlgdrSC0NLD-HX1VU
> I think giving the impact - it would be great to get this approved, and discussed in the TOC as well, since the impact is breaking anyone who depended on the API ( if we are removing gogo generation) ?

Not an unreasonable request, especially given the large nature of the work. I will work with @howardjohn if he is willing on the RFC including the design for gogo removal as he has done most of the RFC work already.

> While it does mot answer your question of whether both types of go bindings can be generated it seems clear that having both types of bindings one based upon protobuf v1 + gogo and another based upon protobuf b2 in the same code baed is asking for pain. We have suffered long enough within Istio with thix problem.
> And we suffered because Envoy decided to do the same thing - pick protobuf and not support gogo, forcing all people depending on Envoy APIs to do the same choice. K8S is still using gogo - and many other projects. Forcing others to follow our choice is simply evil, and if envoy did it we should not follow the same path and inflict pain on others. I'm not saying we shouldn't use the protobuf version - in some cases it'll help, but this is equivalent with an intentional breaking of an API.
> I can look into generate bindings for individuals that wish to stsy on protobuf v1 + gogo which is dead, although my primary motivation is moving us into thd future.

I disagree with your assertion around Envoy.

While some in the community may want to drop gogo as a result of the recent change to go-control-plane, Others have wanted to drop gogo since prior to any action by Envoy. For reference, see [March 2019](https://github.com/istio/operator/pull/382) where I submitted a PR to remove gogo from the operator. @ostromart at the the time convinced me it was not the best course of action for the operator. @geeknoid had spent alot of time trying to move off of gogo, but was stuck in the land of needing to support mixer - which had a legitimate performance need for mixer.

In my initial work, I have found that there is may be no path forward to "protobuf v2" (which is not actually called protobuf v2, but you probably know more of the details). With protobuf 1.4.2, gogo *appears* incompatible.

I am not completely opposed to maintaining a binding for gogo users, although I have to wonder how we would test such a binding if the project were protobuf native (plus enough work to support K8s's use of gogo). I would like to convert the code base away from gogo as much as possible. You can see my [initial work](https://github.com/istio/istio/pull/26769) - which isn't great, but I've only spent about 30 minutes on istio/istio thus far. The other three PRs for the ancillary repositories (including API) with my DNM PRs are in pretty good shape for a gogo eject. I do recognize we need to maintain support for reading K8s objects that use gogo. In these situations, we will convert gogo to protobuf internally, and cache the protobuf versions - where caching is used.

I defer to the experts around the API repository if a version change is needed. While I understand how to use proto from a developer perspective, I don't understand the internal mechanics of wire compatibility. Certainly, ABI compatibility may be broken, and I think that may be ok. Individuals should be using istio/go-client if they are using Istio APIs - this would protect them from such implementation details. If someone is relying on internal implementation details - they are not using a stable API - this raises the question - is a gogo API stable at all - as gogo is not maintained.

> It'll be dead when K8S stops using it.

A project is dead when it is unmaintained and the maintainers step back to do other work. I have to imagine or hope K8s has an exit plan for gogo in their planning. I could go on a long rant about how forking protobuf to produce gogo was the wrong move, but I don't know the political nature of the problem, so I won't say much, other than it would have been better if the gogo developers worked within the protobuf community to produce a better protobuf so we would not be in the situation we are in.

Cheers,
-steve

**Comment by sdake** [comment]
https://github.com/istio/istio/issues/26010

**Comment by costinm** [comment]
On Mon, Aug 24, 2020 at 10:50 PM Steven Dake <notifications@github.com>
wrote:

> On Mon, Aug 24, 2020 at 7:42 PM Steven Dake *@*.***> wrote: @costin
> <https://github.com/costin> https://github.com/costin @howardjohn
> <https://github.com/howardjohn> https://github.com/howardjohn has written
> this excellent document which explains the rationale for moving to gogo:
> https://docs.google.com/document/d/1cyANUWgTP9K4UEmKoFQk_eDR2qTlgdrSC0NLD-HX1VU
> I think giving the impact - it would be great to get this approved, and
> discussed in the TOC as well, since the impact is breaking anyone who
> depended on the API ( if we are removing gogo generation) ?
>
> Not an unreasonable request, especially given the large nature of the
> work. I will work with @howardjohn <https://github.com/howardjohn> if he
> is willing on the RFC including the design for gogo removal as he has done
> most of the RFC work already.
>
> While it does mot answer your question of whether both types of go
> bindings can be generated it seems clear that having both types of bindings
> one based upon protobuf v1 + gogo and another based upon protobuf b2 in the
> same code baed is asking for pain. We have suffered long enough within
> Istio with thix problem.
> And we suffered because Envoy decided to do the same thing - pick protobuf
> and not support gogo, forcing all people depending on Envoy APIs to do the
> same choice. K8S is still using gogo - and many other projects. Forcing
> others to follow our choice is simply evil, and if envoy did it we should
> not follow the same path and inflict pain on others. I'm not saying we
> shouldn't use the protobuf version - in some cases it'll help, but this is
> equivalent with an intentional breaking of an API.
> I can look into generate bindings for individuals that wish to stsy on
> protobuf v1 + gogo which is dead, although my primary motivation is moving
> us into thd future.
>
> I disagree with your assertion around Envoy.
>
> While some in the community may want to drop gogo as a result of the
> recent change to go-control-plane, Others have wanted to drop gogo since
> prior to any action by Envoy. For reference, see March 2019
> <https://github.com/istio/operator/pull/382> where I submitted a PR to
> remove gogo from the operator. @ostromart <https://github.com/ostromart>
> at the the time convinced me it was not the best course of action for the
> operator. @geeknoid <https://github.com/geeknoid> had spent alot of time
> trying to move off of gogo, but was stuck in the land of needing to support
> mixer - which had a legitimate performance need for mixer.
>
> In my initial work, I have found that there is may be no path forward to
> "protobuf v2" (which is not actually called protobuf v2, but you probably
> know more of the details). With protobuf 1.4.2, gogo *appears*
> incompatible.
>
> I am not completely opposed to maintaining a binding for gogo users,
> although I have to wonder how we would test such a binding if the project
> were protobuf native (plus enough work to support K8s's use of gogo). I
> would like to convert the code base away from gogo as much as possible. You
> can see my initial work <https://github.com/istio/istio/pull/26769> -
> which isn't great, but I've only spent about 30 minutes on istio/istio thus
> far. The other three PRs for the ancillary repositories (including API)
> with my DNM PRs are in pretty good shape for a gogo eject. I do recognize
> we need to maintain support for reading K8s objects that use gogo. In these
> situations, we will convert gogo to protobuf internally, and cache the
> protobuf versions - where caching is used.
>
> I defer to the experts around the API repository if a version change is
> needed. While I understand how to use proto from a developer perspective, I
> don't understand the internal mechanics of wire compatibility. Certainly,
> ABI compatibility may be broken, and I think that may be ok. Individuals
> should be using istio/go-client if they are using Istio APIs - this would
> protect them from such implementation details. If someone is relying on
> internal implementation details - they are not using a stable API - this
> raises the question - is a gogo API stable at all - as gogo is not
> maintained.
>
> It'll be dead when K8S stops using it.
>
> A project is dead when it is unmaintained and the maintainers step back to
> do other work. I have to imagine or hope K8s has an exit plan for gogo in
> their planning. I could go on a long rant about how forking protobuf to
> produce gogo was the wrong move, but I don't know the political nature of
> the problem, so I won't say much, other than it would have been better if
> the gogo developers worked within the protobuf community to produce a
> better protobuf so we would not be in the situation we are in.
>

- https://github.com/alexshtin/proto-bench  - it seems even as a zombie,
gogo is still faster ( and remains easier to use ).

- https://github.com/gogo/protobuf/issues/691 - maintainers moving on
doesn't mean a project is dead. Plenty of projects (all apache)
have the list of maintainers change over time, even Istio.

- I'm not sure what you want to test for the gogo bindings. That they are
generated ? That they generated the same bytes as go ? What do you
test for the go protos ?

- I am actually ok with istio/istio moving away from gogo ( or better -
using the new API that allows registry to work properly ). My concern is on
imposing
our preference on users - and it cuts both ways, i.e. I'm 100% for
generating goproto for users who don't like gogo, we should have done it
long ago.

- not clear on what you disagree about my assertion on Envoy. They did make
a choice and imposed their choice on their users, forcing direct and
indirect work on their users. Istio was a gogo-only project before we were
forced by Envoy to pick a dependency they had no need to impose on others,
I'm just saying we shouldn't do the same.

**Comment by kyessenov** [comment]
FYI envoy switched to golang protobuf after soliciting help with gogo from the community and getting no response. There is still a gogo branch available. Anyone is welcome to bring it up to date and improve tooling with protoc-gen-validate, etc.

The runtime performance of protobuf is not known to be poor, regardless of which runtime is used. Protobuf is  optimized for wire serialization, not in-memory data structures.

**Comment by costinm** [comment]
https://github.com/alecthomas/go_serialization_benchmarks - that's
comparing all serializaations for go, for 'wire serialization'.

What I'm asking is to not repeat what envoy did - i.e. force all users of
Istio API to switch to and use only a particular implementation.
We have this discussion in large part because Envoy forced us to use 2
protobuf implementations. Of course, K8S does
the same by using gogo - so at least we should do the right thing.

**Comment by sdake** [comment]
@costinm this touches atleast 6 repositories, so its a bit challenging to even get a proper compile. That said, I am close to measuring the benchmark performance of xds. If the benchmark performance of XDS is not materially different, I am not sure I understand the concern.

As we don't know the performance impact of this change until its actually done, we may end in a situation where we decide not to proceed with this work.

I want to understand if you are also suggesting a gogo binding? After this work, gogo won't work natively in Istio, and I'm not really sure how to do that work without a bunch of conditional logic in the pilot dir.

Please take a quick look at what I've been doing if you can. I don't mind investing the time even if this work doesn't merge. But I'd like your input if I am on the right track.

Cheers,
-steve

PR: https://github.com/istio/istio/issues/26010
PR: https://github.com/istio/tools/pull/1183
PR: https://github.com/istio/common-files/pull/303
PR: https://github.com/istio/istio/pull/26769
PR: https://github.com/istio/api/pull/1607

**Comment by sdake** [comment]
Zombie PR. Closing.

**Pull request closed without merging**